R/new-blog-post.R

```diff
@@ -0,0 +1,82 @@
+#' Create a new blog post
+#'
+#' Creates (and potentially opens) the `index.qmd` file for a new blog post.
+#'
+#' @param title A character string for the title of the post. It is converted
+#' to title case via [tools::toTitleCase()].
+#' @param dest A character string (or NULL) for the path within `posts`. By
+#' default, the title is adapted as the directory name.
+#' @param open A logical: have the default editor open a window to edit the
+#' `index.qmd` file?
+#' @param call A call object for reporting errors.
+#' @param ... A named list of values to be added to the yaml header, such as
+#' `description`, `author`, `categories`, etc.
+#' @return The path to the index file.
+#' @export
+#' @examples
+#' \dontrun{
+#'  \donttest{
+#' new_blog_post("making quarto blog posts", categories = c("R"))
+#'
+#'  }
+#' }
+#'
+new_blog_post <- function(title, dest = NULL, open = rlang::is_interactive(),
+                          call = rlang::current_env(), ...) {
+
+  if (is.null(dest)) {
+    # Scrub title to make directory name
+    dest <- gsub("[[:space:]]", "-", tolower(title))
+  }
+  dest_path <- make_post_dir(dest, call)
+  post_yaml <- make_post_yaml(title, ...)
+  qmd_path <- write_post_yaml(post_yaml, dest_path, call)
+  if (open) {
+    utils::file.edit(qmd_path)
+  }
+  invisible(qmd_path)
+}
+
+make_post_dir <- function(dest, call) {
+  working <- fs::path_wd()
+
+  post_path <- fs::path(working, "posts", dest)
+
+  if (fs::dir_exists(post_path)) {
+    cli::cli_abort("There is already a {.code {dest}} directory in 'posts/'",
+                   call = call)
+  } else {
+    ret <- fs::dir_create(post_path)
+  }
+  ret
+}
+
+make_post_yaml <- function(title, ...) {
+  default_values <- list(
+    title = tools::toTitleCase(title),
+    author = tools::toTitleCase(whoami::fullname("Your name")),
+    date = format(Sys.Date(), "%Y-%m-%d"),
+    categories = character(0)
+  )
+
+  user_values <- list(...)
+
+  yml_values <- utils::modifyList(default_values, user_values)
+  if (length(yml_values$categories) == 0) {
+    yml_values <- yml_values[names(yml_values) != "categories"]
+  }
+  yml_values <- yaml::as.yaml(yml_values)
+  yml_values <- paste0("---\n", yml_values, "---\n")
```

Comment:
Also, maybe there is a better API.

Reply:
While dealing with YAML in quarto context we have the issue of **yaml** R package being 1.1 spec, and Quarto using 1.2 spec. 

So it requires some specific handlers as workaround. See 
https://github.com/quarto-dev/quarto-r/blob/d71b135768cda83a3438e22362115ed7c2bbabd6/R/utils.R#L6-L16

We probably need to adapt for generating not in a file, or refactor the handlers so that we can use them in other functions. 


Regaring `---`, I believe this is good enough. 

This post creation may also be done through templating 🤔 Having a whisker template (or else) to file out with some information. Probably more complex than necessary, but we did that for bookdown skeleton for example using placeholder in template file (https://github.com/rstudio/bookdown/blob/f244cf12bf2c2d7106ac6322b2b2a5796d4ef0c8/R/skeleton.R#L66-L83) 

I am fine with current way